dspace-api/src/main/java/org/dspace/curate/QueueManager.java

```diff
@@ -0,0 +1,102 @@
+import org.dspace.core.Context;
+import org.dspace.services.ConfigurationService;
+import org.dspace.services.factory.DSpaceServicesFactory;
+
+/**
+ * Helper class to deal with curation tasks.
+ *
+ * @author Andrea Schweer schweer@waikato.ac.nz for the LCoNZ IRRs
+ * @author Kim Shepherd kim@shepherd.nz update and refactor for DSpace 6.x generic usage
+ */
+public class QueueManager {
+    private static final Logger log = Logger.getLogger(QueueManager.class);
+    private static final ConfigurationService configurationService =
+            DSpaceServicesFactory.getInstance().getConfigurationService();
+
+    private final List<String> taskNames = new ArrayList<String>();
+    private String queueName = "continually";
+    private ArrayList<Item> toQueue;
+
+    /**
+     * Retrieve queue name from configuration, for a given event consumer
+     * @param queueProperty the configuration property string returned by getQueueProperty() in the consumer itself
+     */
+    public void initQueueName(String queueProperty) {
+        String queueConfig = configurationService.getProperty(queueProperty);
+        if (queueConfig != null && !"".equals(queueConfig)) {
+            queueName = queueConfig;
+            log.info("Using queue name " + queueName);
+        } else {
+            log.info("No queue name specified, using default: " + queueName);
+        }
+    }
+
+    /**
+     * Retrieve task names from configuration, for a given event consumer
+     * @param tasksProperty the configuration property string returned by getTaskProperty() in the consumer itself
+     */
+    public void initTaskNames(String tasksProperty) {
+        String[] taskConfig = configurationService.getArrayProperty(tasksProperty, new String[]{});
+        taskNames.addAll(Arrays.asList(taskConfig));
+        log.info("Setting up tasks as " + Arrays.deepToString(taskNames.toArray()));
+    }
+
+    /**
+     * Add a resolved item to the list of items to queue (once the consumer has finished all other processing)
+     * @param item Dspace item to queue
+     */
+    public void addToQueue(Item item) {
+        if (toQueue == null) {
+            toQueue = new ArrayList<Item>();
+        }
+        toQueue.add(item);
+        log.info("Adding item " + item.getHandle() + " to list of items to queue");
+    }
+
+    /**
+     * Take the resolved toQueue items and queue them for curation in the configured queue name
+     * with the configured task name(s). This is called at the end of the consumer processing.
+     * @param ctx DSpace context
+     * @throws IOException
+     */
+    public void queueForCuration(Context ctx) throws IOException {
+        if (toQueue != null && !toQueue.isEmpty()) {
+            log.info("Actually queueing " + toQueue.size() + " items for curation");
+            for (String taskName : taskNames) {
+                Curator curator = new Curator().addTask(taskName);
+                for (Item item : toQueue) {
```

Comment:
Why do we create a separate Curator for each task, rather than add all of the tasks to a single Curator and then queue each Item once for the whole list of tasks?

Reply:
My understanding here is to enforce the order of queued tasks, in the `addTask()` comments, it says:

> Caller should make no assumptions on execution ordering.

When queuing tasks on item install, for instance, some tasks might need to happen before others. Happy to take a closer look though, because perhaps it's possible to alter the Curator task handling so we **can** enforce order there.

Reply:
The tasks are currently added to a List and run in list order.  See the `curate` methods.  If this is generally useful behavior, maybe execution in order of addition should be made part of `addTask`'s contract.  The comment dates from 2010, as does this one:  "// performance order currently FIFO - to be revisited".  No revisitation has occurred.

I think that undefined execution order violates the principle of least surprise and should not be in the contract.  But this is really a separate issue.  OTOH changing the *stated* behavior requires no changes to actual code.  I will leave the choice to you.  I simply wanted to note that this seemed inefficient.  Thanks for noticing that comment.